### DISCUSSION TITLE
Second layer API fetch / Toggle logic to show/hide data

### DISCUSSION THREAD

**Discussion by StefanieLoeser** [neuefische/questions#218] [Web] [answered]
Hi everybody,
I made a fetch to second layer of API and want to show/hide these data via a toggle button.
The fetch created the following error that I tried to fix by adding 'Access-Control-Allow-Origin' to the fetch header.
It seemed to work at first, but now the toggle doesn't always show the hidden information. 
Thank you for your help!

[Link to PR](https://github.com/StefanieLoeser/capstone-project/pull/17)

**error**

![Bildschirmfoto 2022-10-05 um 15 47 11](https://user-images.githubusercontent.com/104068626/194076318-124d805a-65bb-40d8-800a-7a4da85d35ca.png)

**fix within first fetch**
```const init = {
    headers: {
      'User-Agent': 'MyRecordsPlaylistApp/1.0.0 +http://localhost:3000/',
      Authorization: 'Discogs token=' + userToken,
      'Access-Control-Allow-Origin': '*',
    },
  };
  const res = await fetch(collectionItemsByFolderURL, init);
  const data = await res.json();
  ```

**second fetch & toggle logic**
```nextjs
export default function RecordFile({ record, onToggleBookmark }) {
  const releaseApiUrl = record.basic_information?.resource_url;
  const [tracklist, setTracklist] = useState([]);

  useEffect(() => {
    fetch(releaseApiUrl)
      .then((response) => response.json())
      .then((data) => setTracklist(data.tracklist))
      .catch((error) => console.error(error));
  }, []);

  const [showTracks, setShowTracks] = useState(false);
  function toggleTracks() {
    setShowTracks(!showTracks);
  }

  return (
    <Record>
      <Cover src={record.basic_information?.thumb} />
      <div>
        <BookmarkIcon
          type="checkbox"
          checked={record.isChecked}
          onChange={onToggleBookmark}
        />
        <RecordDetails>
          <li key={record.id}>
            <strong>{record.basic_information?.artists[0].name}</strong>
          </li>
          <li key={record.instance_id}>
            <em>
              <strong>{record.basic_information?.title}</strong>
            </em>
            , {record.basic_information?.year}
          </li>
          <li key={record.id}>{record.basic_information?.labels[0].name}</li>
        </RecordDetails>
        <TrackInformation
          showTracks={showTracks}
          style={{ display: showTracks ? 'block' : 'none' }}
        >
          {tracklist.map((track) => {
            return (
              <Tracks key={track.position + track.title}>
                {track.position}-
                <strong>
                  {track.artists ? `${track.artists[0].name} -` : ''}
                </strong>
                <em>{track.title}</em>{' '}
                {track.duration ? `(${track.duration} min)` : ''}
              </Tracks>
            );
          })}
        </TrackInformation>
        <ButtonIcon onClick={toggleTracks}>{showTracks ? 'x' : '+'}</ButtonIcon>
      </div>
    </Record>
  );
}

```

**Comment by ThomasKirst** [comment, marked as answer]
Hi Stefanie,

are the HTTP Headers from your first API fetch missing in the second one – especially this part:
`Authorization: 'Discogs token=' + userToken`?

When you have a look at your network inspector you will see that some request pass and others are failing:

![Screenshot 2022-10-05 at 16 25 11](https://user-images.githubusercontent.com/4458383/194086160-d7cc6246-2a0d-4826-9bd9-8536cace8d0f.png)

The status code 429 means `too many requests`:
<img src="http://http.cat/429" width="400" />

The HTTP header ` 'Access-Control-Allow-Origin': '*'` is one which is sent by the server. It is not sent by us from the client against the API.

Here you see the response headers:

![Screenshot 2022-10-05 at 16 26 11](https://user-images.githubusercontent.com/4458383/194086917-dc02adbf-027d-410e-9ce3-0ffc6b9e113a.png)

Maybe it´s already sufficient to make the calls with the API token.

The API token might be a secret that is not supposed to be used in the client code.
For a thorough solution you will probably need an API proxy (Client API call –> NextJS API handler() –> Discogs API).

– Thomas

**Reply by StefanieLoeser** [reply]
Thank you @ThomasKirst! 

The strange thing is, that it worked for all fetched records without the token before I added the toggle logic to hide or show the data. 

I tried to add the token to the second fetch, also using the consumer key + secret instead –  but no success either way. Also I thought the API endpoint might be accessible without authentication as the provided API routes from the first fetch can simply be clicked and show the data behind (e.g. https://api.discogs.com/releases/594348). 

Rereading the API documentation unfortunately didn't clear things up either. 
Here is the [docu on fetching releases](https://www.discogs.com/developers#page:database,header:database-release-get), where the general GET release doesn't seem to need authentication. (It would say otherwise).

<img width="776" alt="Bildschirmfoto 2022-10-05 um 21 30 42" src="https://user-images.githubusercontent.com/104068626/194146265-e24deb54-9657-4ff5-9128-57c227b7410b.png">

_--> Here it says auth. is needed:_

<img width="758" alt="Bildschirmfoto 2022-10-05 um 21 33 01" src="https://user-images.githubusercontent.com/104068626/194146675-5b373e6e-7d55-4177-9f54-8bd5ad6b3042.png">

Do you see any other possible way to make it work without an API proxy (which seems quite complex)?

Thank you, Stefanie

**Reply by StefanieLoeser** [reply]
Thank you @ThomasKirst for your Zoom support!
Moving the fetch into the toggle logic did the trick. 💯

**Reply by ThomasKirst** [reply]
Could you please provide the code snippet(s) that solved your problem here and mark the question as answered?
Thank you :)

**Reply by StefanieLoeser** [reply]
Sure thing!

```
export default function RecordFile({ record, onToggleBookmark }) {
  const releaseApiUrl = record.basic_information?.resource_url;
  const [tracklist, setTracklist] = useState([]);
  const [showTracks, setShowTracks] = useState(false);

  function toggleTracks() {
    setShowTracks(!showTracks);
    fetch(releaseApiUrl)
      .then((response) => response.json())
      .then((data) => setTracklist(data.tracklist))
      .catch((error) => console.error(error));
  }

  return (
    <Record>
      <Cover src={record.basic_information?.thumb} />
      <div>
        <BookmarkIcon
          type="checkbox"
          checked={record.isChecked}
          onChange={onToggleBookmark}
        />
        <RecordDetails>
          <li key={`#${record.id}`}>
            <strong>{record.basic_information?.artists[0].name}</strong>
          </li>
          <li key={record.instance_id}>
            <em>
              <strong>{record.basic_information?.title}</strong>
            </em>
            , {record.basic_information?.year}
          </li>
          <li key={record.id}>{record.basic_information?.labels[0].name}</li>
        </RecordDetails>
        <TrackInformation
          showTracks={showTracks}
          style={{ display: showTracks ? 'block' : 'none' }}
        >
          {tracklist.map((track) => {
            return (
              <Track key={track.position + track.title}>
                {track.position}-
                <strong>
                  {track.artists ? `${track.artists[0].name} - ` : ''}
                </strong>
                <em>{track.title}</em>{' '}
                {track.duration ? `(${track.duration} min)` : ''}
              </Track>
            );
          })}
        </TrackInformation>
        <ButtonIcon onClick={toggleTracks}>{showTracks ? '-' : '+'}</ButtonIcon>
      </div>
    </Record>
  );
}
```
